test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs

```diff
@@ -0,0 +1,108 @@
+using Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests;
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.AspNetCore.Testing.xunit;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.Logging;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace IISIntegration.FunctionalTests.Inprocess
+{
+    [SkipIfIISExpressSchemaMissingInProcess]
+    public class AppOfflineTests : IISFunctionalTestBase
+    {
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteIsDown_SiteReturns503()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteRunning_SiteShutsDown()
+        {
+            var deploymentResult = await StartSite();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            while (response.IsSuccessStatusCode)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+            }
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
+
+            deploymentResult.DeploymentResult.HostShutdownToken.WaitHandle.WaitOne(millisecondsTimeout: 1000);
+
+            Assert.True(deploymentResult.DeploymentResult.HostShutdownToken.IsCancellationRequested);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDropped_CanRemoveAppOfflineAfterAdding()
+        {
+            var deploymentResult = await StartSite();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            while (response.IsSuccessStatusCode)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+            }
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
+
+            RemoveAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+        }
+
+        private async Task<IISDeploymentResult> StartSite()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
```

Comment:
Too many hello worlds around. Let's add VerifyStarted helper and use it everywhere.